src/tools/wasm-fuzz-lattices.cpp

```diff
@@ -158,39 +158,47 @@ using ArrayLattice = analysis::Array<RandomLattice, 2>;
 using TupleFullLattice = analysis::Tuple<RandomFullLattice, RandomFullLattice>;
 using TupleLattice = analysis::Tuple<RandomLattice, RandomLattice>;
 
-struct RandomFullLattice::L : std::variant<Bool,
-                                           UInt32,
-                                           ValType,
-                                           Inverted<RandomFullLattice>,
-                                           ArrayFullLattice,
-                                           Vector<RandomFullLattice>,
-                                           TupleFullLattice> {};
-
-struct RandomFullLattice::ElementImpl
-  : std::variant<typename Bool::Element,
-                 typename UInt32::Element,
-                 typename ValType::Element,
-                 typename Inverted<RandomFullLattice>::Element,
-                 typename ArrayFullLattice::Element,
-                 typename Vector<RandomFullLattice>::Element,
-                 typename TupleFullLattice::Element> {};
-
-struct RandomLattice::L : std::variant<RandomFullLattice,
-                                       Flat<uint32_t>,
-                                       Lift<RandomLattice>,
-                                       ArrayLattice,
-                                       Vector<RandomLattice>,
-                                       TupleLattice,
-                                       Shared<RandomLattice>> {};
-
-struct RandomLattice::ElementImpl
-  : std::variant<typename RandomFullLattice::Element,
-                 typename Flat<uint32_t>::Element,
-                 typename Lift<RandomLattice>::Element,
-                 typename ArrayLattice::Element,
-                 typename Vector<RandomLattice>::Element,
-                 typename TupleLattice::Element,
-                 typename Shared<RandomLattice>::Element> {};
+using FullLatticeVariant = std::variant<Bool,
+                                        UInt32,
+                                        ValType,
+                                        Inverted<RandomFullLattice>,
+                                        ArrayFullLattice,
+                                        Vector<RandomFullLattice>,
+                                        TupleFullLattice>;
+
+struct RandomFullLattice::LatticeImpl : FullLatticeVariant {};
```

Comment:
Is this "subclass without adding anything" a workaround for a compiler bug, or a C++ spec requirement, or something else?

Reply:
Even before this PR, we had to subclass without adding anything because we had to defer defining these nested types until their outer types (`RandomFullLattice` and `RandomLattice`) were complete. Since `using` statements cannot be forward declared, we had no choice but to forward declare `LatticeImpl` (previously `L`) as separate structs.

This PR just gives a name to the raw `std::variant<...>` superclasses as well and uses those newly-named types down below where we call `std::visit`.